esinstall/src/entrypoints.ts

```diff
@@ -0,0 +1,268 @@
+import {ExportField, ExportMapEntry, PackageManifestWithExports, PackageManifest} from './types';
+import {parsePackageImportSpecifier, resolveDependencyManifest} from './util';
+
+// Rarely, a package will ship a broken "browser" package.json entrypoint.
+// Ignore the "browser" entrypoint in those packages.
+const BROKEN_BROWSER_ENTRYPOINT = ['@sheerun/mutationobserver-shim'];
+
+function hasBrokenBrowserEntrypoint(packageName?) {
+  return !!(packageName && BROKEN_BROWSER_ENTRYPOINT.includes(packageName));
+}
+
+function manifestHasTypes(manifest: PackageManifest): boolean {
+  return !!(manifest.types || manifest.typings);
+}
+
+type ResolveManifestEntryOptions = {
+  packageLookupFields?: string[];
+  packageName?: string;
+};
+
+function resolveManifestEntry(
+  manifest: PackageManifest,
+  entry?: string,
+  {packageLookupFields = [], packageName}: ResolveManifestEntryOptions = {},
+): string | undefined {
+  let foundEntrypoint: string | undefined;
+
+  if (manifest.exports) {
+    foundEntrypoint =
+      typeof manifest.exports === 'string'
+        ? manifest.exports
+        : resolveExportMapEntry(manifest.exports['.']);
+
+    if (typeof foundEntrypoint === 'string') {
+      return foundEntrypoint;
+    }
+  }
+
+  foundEntrypoint = [
+    ...packageLookupFields,
+    'browser:module',
+    'module',
+    'main:esnext',
+    'jsnext:main',
+  ]
+    .map((e) => manifest[e])
+    .find(Boolean);
+
+  if (!foundEntrypoint && !hasBrokenBrowserEntrypoint(packageName)) {
+    // Some packages define "browser" as an object. We'll do our best to find the
+    // right entrypoint in an entrypoint object, or fail otherwise.
+    // See: https://github.com/defunctzombie/package-browser-field-spec
+    let browserField = manifest.browser;
+
+    if (typeof browserField === 'string') {
+      foundEntrypoint = browserField;
+    } else if (typeof browserField === 'object') {
+      let browserEntrypoint =
+        (entry && browserField[entry]) ||
+        browserField['./index.js'] ||
+        browserField['./index'] ||
+        browserField['index.js'] ||
+        browserField['index'] ||
+        browserField['./'] ||
+        browserField['.'];
+
+      if (typeof browserEntrypoint === 'string') {
+        foundEntrypoint = browserEntrypoint;
+      }
+    }
+  }
+
+  // If browser object is set but no relevant entrypoint is found, fall back to "main".
+  if (!foundEntrypoint) {
+    foundEntrypoint = manifest.main;
+  }
+
+  return foundEntrypoint;
+}
+
+export function resolveMainEntrypoint(manifest: PackageManifest) {
+  return resolveManifestEntry(manifest);
+}
+
+/**
+ * Given an ExportMapEntry find the entry point, resolving recursively.
+ */
+export function resolveExportMapEntry(
+  exportMapEntry?: ExportMapEntry,
+  conditions?: string[],
+): string | undefined {
+  // If this is a string or undefined we can skip checking for conditions
+  switch (typeof exportMapEntry) {
```

Comment:
nit: slight preference for `if (!exportMapEntry || typeofExportMapEntry === 'string') {`